### PR TITLE
New markdown editor

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Implements #11240.

https://github.com/user-attachments/assets/4d2f8021-3e0f-4d39-95df-bcd72bf7545b

### Important Notes

- Fix a Yjs document corruption bug caused by `DeepReadonly` being replaced by a stub; introduce a `DeepReadonly` implementation without Vue dependency.
- Fix right panel sizing when code editor is open.
- Fix right panel slide-in animation.
- `Ast.Function` renamed to `Ast.FunctionDef`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

File: app/common/src/utilities/data/iterator.ts
Comment (farmaazon) [inline app/common/src/utilities/data/iterator.ts:1]: Let's merge in `app/gui/src/project-view/util/iterable.ts` to have one place with itera(tor|ble) utilities. Also, `map` is duplicated there.

File: app/gui/src/project-view/components/DockPanel.vue [hunk 1]
```diff
@@ -90,11 +94,23 @@ const tabStyle = {
...
+  width: var(--dock-panel-width, var(--right-dock-default-width));
 }
 
-.DockPanelContent {
-  min-width: 258px;
-  width: var(--dock-panel-width);
+.panelInner {
+  min-width: var(--dock-panel-min-width);
+  /* FIXME: Need a fixed size when `dock-panel-width` is not set, so that animation works for default-sized panels. */
+  width: var(--dock-panel-width, 100%);
```
Comment (farmaazon) [inline app/gui/src/project-view/components/DockPanel.vue]: What is here to be fixed? What hold us from fixing it now?

File: app/gui/src/project-view/components/GraphEditor.vue [hunk 2]
```diff
@@ -43,22 +42,24 @@ import { provideInteractionHandler } from '@/providers/interactionHandler'
 import { provideKeyboard } from '@/providers/keyboard'
 import { injectVisibility } from '@/providers/visibility'
 import { provideWidgetRegistry } from '@/providers/widgetRegistry'
-import { provideGraphStore, type NodeId } from '@/stores/graph'
-import type { RequiredImport } from '@/stores/graph/imports'
+import { NodeId, provideGraphStore } from '@/stores/graph'
+import { RequiredImport } from '@/stores/graph/imports'
```
Comment (farmaazon) [inline app/gui/src/project-view/components/GraphEditor.vue]: Why this change? We still use `RequiredImport` as type only. There are many similar changes in imports here.

File: app/gui/src/project-view/components/MarkdownEditor/EditorRoot.vue
Comment (farmaazon) [inline app/gui/src/project-view/components/MarkdownEditor/EditorRoot.vue:1]: What is the responsibility of this component? When adding anything new, what should go here, and what to `MarkdownEditorImpl`?

File: app/gui/src/project-view/components/MarkdownEditor/MarkdownEditorImpl.vue [hunk 3]
```diff
@@ -1,80 +1,137 @@
...
+import { ComponentInstance, onMounted, ref, toRef, useCssModule, watchEffect } from 'vue'
+import { yCollab } from 'y-codemirror.next'
+import * as awarenessProtocol from 'y-protocols/awareness.js'
+import * as Y from 'yjs'
+
+const editorRoot = ref<ComponentInstance<typeof EditorRoot>>()
 
-const markdown = defineModel<string>({ required: true })
 const props = defineProps<{
+  yText: Y.Text
```
Comment (farmaazon) [inline app/gui/src/project-view/components/MarkdownEditor/MarkdownEditorImpl.vue:21]: Having `Y.Text` as a prop makes it deeply reactive. Now, to some extent this is desirable (so we don't need to listen to Ydoc updates), but I wonder if there are any guarantees what's inside Y.Text object? I'd like to avoid a situation when this property is updated because something changed in a "global" Ydoc structure, which is back referenced here.

Or prehaps it's enough to have it shallow-reactive? I don't know how yCollab exactly works.

File: app/gui/src/project-view/components/MarkdownEditor/markdown/lezer.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+  export interface Buffer {
+    content: number[]
+    nodes: Tree[]
+
+    write: (type: Type, from: number, to: number, children?: number) => Buffer
+    writeElements: (elts: readonly (Element | TreeElement)[], offset?: number) => Buffer
+    finish: (type: Type, length: number) => Tree
+  }
+}
```
Comment (farmaazon) [inline app/gui/src/project-view/components/MarkdownEditor/markdown/lezer.ts]: Shouldn't we keep these in some d.ts?

File: app/gui/src/project-view/util/ast/abstract.ts [hunk 5]
```diff
@@ -271,6 +278,79 @@ export function dropMutability<T extends Ast>(value: Owned<Mutable<T>>): T {
...
+function unwrapGroups(ast: Ast) {
+  while (ast instanceof Group && ast.expression) ast = ast.expression
+  return ast
+}
+
+/**
+ * Tries to recognize inputs that are semantically-equivalent to a sequence of `App`s, and returns the arguments
+ *  identified and LHS of the analyzable chain.
+ *
+ *  In particular, this function currently recognizes syntax used in visualization-preprocessor expressions.
```
Comment (farmaazon) [inline app/gui/src/project-view/util/ast/abstract.ts]:
```suggestion
 * Tries to recognize inputs that are semantically-equivalent to a sequence of `App`s, and returns the arguments
 * identified and LHS of the analyzable chain.
 *
 * In particular, this function currently recognizes syntax used in visualization-preprocessor expressions.
```

File: app/ydoc-shared/src/ast/mutableModule.ts [hunk 6]
```diff
@@ -272,23 +271,28 @@ export class MutableModule implements Module {
...
-        updateBuilder.updateWidgets(id)
+        if (event.target === metadata) {
+          const changes: (readonly [string, unknown])[] = Array.from(
+            event.changes.keys,
+            ([key]) => [key, metadata.get(key as any)],
+          )
+          updateBuilder.updateMetadata(id, changes)
+        } else if (event.target.parent === metadata) {
+          // Updates to some specific widget's metadata
+          const id = event.target.parent.parent.get('id')
```
Comment (farmaazon) [inline app/ydoc-shared/src/ast/mutableModule.ts]: If we needed to add a check above, that `event.target` is metadata, then maybe we should also check here, that event.target is `widget` of `metadata`?

File: app/ydoc-shared/src/ast/tree.ts [hunk 7]
```diff
@@ -139,12 +197,12 @@ export abstract class Ast {
     return metadata as FixedMapView<NodeMetadataFields & MetadataFields>
   }
 
-  /** Get metadata of all widgets assigned to this node. */
-  widgetsMetadata(): FixedMapView<Record<string, unknown>> {
+  /** Get metadata of all widgets assigned to this node. WARNING: The returned value must not be mutated. */
+  widgetsMetadata(): ReadonlyMap<string, DeepReadonly<unknown>> {
```
Comment (farmaazon) [inline app/ydoc-shared/src/ast/tree.ts:199]: On this occasion, add docs for `nodeMetadata` above (with similar warning if needed).

File: app/gui/src/project-view/components/GraphEditor.vue (hunk 2 above)
Comment (kazcw) [reply app/gui/src/project-view/components/GraphEditor.vue]: Didn't we used to have a lint for that? But we don't now, and it seems `consistent-type-imports` isn't compatible with Vue parser...

File: app/ydoc-shared/src/ast/tree.ts (hunk 7 above)
Comment (kazcw) [reply app/ydoc-shared/src/ast/tree.ts:199]: Added docs. `nodeMetadata` is safe, we just can't enforce readonly access to `widgetsMetadata` because it doesn't have a known type to apply the `DeepReadonly` transformation to.

File: app/common/src/utilities/data/iterator.ts
Comment (kazcw) [reply app/common/src/utilities/data/iterator.ts:1]: Made a unified iterables library, and added some tests for it.

File: app/gui/src/project-view/components/MarkdownEditor/EditorRoot.vue
Comment (kazcw) [reply app/gui/src/project-view/components/MarkdownEditor/EditorRoot.vue:1]: `EditorRoot` is the non markdown/documentation-specific base container. Moved it to `components` and refactored `CodeEditor` to use it too.

File: app/gui/src/project-view/components/MarkdownEditor/MarkdownEditorImpl.vue (hunk 3 above)
Comment (kazcw) [reply app/gui/src/project-view/components/MarkdownEditor/MarkdownEditorImpl.vue:21]: Good catch--yCollab listens for Yjs updates, so we don't need deep reactivity here. Refactored this component to watch the `yText` shallowly.

Generally speaking, Yjs objects are reasonably well-behaved if deep reactivity is applied to them. When a Yjs object is watched deeply it shouldn't cause *completely* spurious invalidation due to activity elsewhere in the document, but some more local spurious invalidation can occur due to mutation of objects that are adjacent in Yjs's internal doubly-linked list. The main reason it's more efficient to use them non-reactively and listen to Yjs updates when possible, is that we can make smaller changes when we receive a Yjs delta than when Vue reactivity tells us we need to recompute something.

File: app/gui/src/project-view/components/ComponentBrowser/component.ts [hunk 8]
```diff
@@ -10,7 +10,8 @@ import { isSome } from '@/util/data/opt'
 import { Range } from '@/util/data/range'
 import { displayedIconOf } from '@/util/getIconName'
 import type { Icon } from '@/util/iconName'
-import { qnLastSegmentIndex, QualifiedName, tryQualifiedName } from '@/util/qualifiedName'
+import type { QualifiedName } from '@/util/qualifiedName'
+import { qnLastSegmentIndex, tryQualifiedName } from '@/util/qualifiedName'
```
Comment (farmaazon) [inline app/gui/src/project-view/components/ComponentBrowser/component.ts:14]: My vscode prettier put those to a single line:
```ts
import { qnLastSegmentIndex, tryQualifiedName, type QualifiedName } from '@/util/qualifiedName'
```

File: app/gui/src/project-view/components/MarkdownEditor/markdown/decoration.ts [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+  doc: Text,
+  emitDecoration: (from: number, to: number, deco: Decoration) => void,
+) {
+  if (nodeRef.name === 'Link') {
+    const node = nodeRef.node
+    const pi0 = node.firstChild // [
+    const pi1 = pi0?.nextSibling // ]
+    const pi2 = pi1?.nextSibling // (
+    const url = pi2?.nextSibling
+    if (!url || !pi0 || !pi1) {
```
Comment (farmaazon) [inline app/gui/src/project-view/components/MarkdownEditor/markdown/decoration.ts]: If I understand the chain above, `url` is not falsy only when pi0 and pi1 are not null (thus not falsy, as we expect objects there).

File: app/gui/src/project-view/components/MarkdownEditor/markdown/decoration.ts [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+function decorateLink(
+  nodeRef: SyntaxNodeRef,
+  doc: Text,
+  emitDecoration: (from: number, to: number, deco: Decoration) => void,
+) {
+  if (nodeRef.name === 'Link') {
+    const node = nodeRef.node
+    const pi0 = node.firstChild // [
+    const pi1 = pi0?.nextSibling // ]
+    const pi2 = pi1?.nextSibling // (
```
Comment (farmaazon) [inline app/gui/src/project-view/components/MarkdownEditor/markdown/decoration.ts]: I don't know what `pi` means. Maybe `text_open`, `text_close`, `url_open` etc.

File: app/gui/src/project-view/components/MarkdownEditor/markdown/decoration.ts [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+function decorateImageWithRendered(
+  nodeRef: SyntaxNodeRef,
+  doc: Text,
+  emitDecoration: (from: number, to: number, deco: Decoration) => void,
+  vueHost: VueHost,
+) {
+  if (nodeRef.name === 'Image') {
+    const node = nodeRef.node
+    const alt = 'Image' // TODO
```
Comment (farmaazon) [inline app/gui/src/project-view/components/MarkdownEditor/markdown/decoration.ts]: TODO left.

**Review by farmaazon — APPROVED** [review]

File: app/gui/src/project-view/components/MarkdownEditor/markdown/decoration.ts (hunk 9 above)
Comment (kazcw) [reply app/gui/src/project-view/components/MarkdownEditor/markdown/decoration.ts]: The typechecker seems unable to narrow the variables transitively. Refactored it to pass typecheck without looking weird to humans.
